### PR TITLE
Fixes 1172: Clear Filter

### PR DESCRIPTION
**Pull request by kwarnerredhat**
## Testing steps-
"clicking clear filters button should remove any filters selected from popular repositories table"

### CODE REVIEW

**Comment by jlsherrill** [comment]
https://issues.redhat.com/browse/HMS-1172

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx [hunk 1]
```diff
@@ -128,6 +137,12 @@ const PopularRepositoriesTable = () => {
   const debouncedSearchValue = useDebounce(searchValue);
   const [perPage, setPerPage] = useState(storedPerPage);
   const [isActionOpen, setIsActionOpen] = useState(false);
+  const [filterData, setFilterData] = useState<FilterData>({
+    searchQuery: '',
+    versions: [],
+    arches: [],
+    statuses: [],
+  });
```
Comment (Andrewgdewar) [inline src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx:145]: this is not necessary, we are only using searchValue.

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx [hunk 2]
```diff
@@ -343,6 +358,10 @@ const PopularRepositoriesTable = () => {
 
   const countIsZero = !data?.data?.length;
 
+  const clearFilters = () =>
+    setFilterData({ searchQuery: '', versions: [], arches: [], statuses: [] });
+  setSearchValue('');
```
Comment (Andrewgdewar) [inline src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx:363]: what does this line of code do.

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx [hunk 3]
```diff
@@ -343,6 +358,10 @@ const PopularRepositoriesTable = () => {
 
   const countIsZero = !data?.data?.length;
 
+  const clearFilters = () =>
+    setFilterData({ searchQuery: '', versions: [], arches: [], statuses: [] });
```
Comment (Andrewgdewar) [inline src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx:362]: Remove this, not needed.

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx [hunk 4]
```diff
@@ -466,6 +485,17 @@ const PopularRepositoriesTable = () => {
               isCompact
               onPerPageSelect={onPerPageSelect}
             />
+            <Hide
+              hide={
+                !debouncedSearchValue &&
+                filterData.versions.length === 0 &&
+                filterData.arches.length === 0
```
Comment (Andrewgdewar) [inline src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx:492]: Remove this, not needed.

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx [hunk 5]
```diff
@@ -597,7 +627,7 @@ const PopularRepositoriesTable = () => {
       </Hide>
       <Hide hide={!countIsZero}>
         <EmptyTableState
-          clearFilters={() => setSearchValue('')}
+          clearFilters={clearFilters}
```
Comment (Andrewgdewar) [inline src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx:630]: undo this change.

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx (hunk 2 above)
Comment (kwarnerredhat) [reply src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx:363]: I added that from the filters on the Content List Filters following the pattern but realized what you told me earlier

Comment (Andrewgdewar) [reply src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx:363]: This line of code is not a part of the "clearFilters" function above. It is separate. 
What does this line of code do by itself here in such a case?

Comment (kwarnerredhat) [reply src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx:363]: I removed the line it worked

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx [hunk 6]
```diff
@@ -452,6 +474,18 @@ const PopularRepositoriesTable = () => {
...
+              <Chip key='search_chip' onClick={() => setSearchQuery('')}>
+                {searchQuery}
+              </Chip>
+            </ChipGroup>
+          )}
+          {(debouncedSearchValue !== '' && searchQuery !== '') || (
+            <Button className={classes.clearFilters} onClick={clearFilters} variant='link' isInline>
+              Clear filters
+            </Button>
+          )}
```
Comment (Andrewgdewar) [inline src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx:488]: This code always shows, you need to hide the clear filters button when no search query is made.

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx [hunk 7]
```diff
@@ -452,6 +474,18 @@ const PopularRepositoriesTable = () => {
               </FlexItem>
             </InputGroupItem>
           </InputGroup>
+          {searchQuery !== '' && (
+            <ChipGroup categoryName='Name/URL'>
+              <Chip key='search_chip' onClick={() => setSearchQuery('')}>
+                {searchQuery}
+              </Chip>
+            </ChipGroup>
```
Comment (Andrewgdewar) [inline src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx:482]: This chip isn't showing yet.

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx [hunk 8]
```diff
@@ -30,54 +30,66 @@ it('expect PopularRepositoriesTable to render with add one item', () => {
     data: { data: [defaultPopularRepository], meta: { count: 1, limit: 20, offset: 0 } },
```
Comment (Andrewgdewar) [inline src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx:30]: In this file, you have changed a lot of code that doesn't need to be change.
Revert all changes, and either: 

A. **Add** your own test

B. **Add** code to an existing test. 

Do not delete/move any code that currently exists here please.

**Comment by Andrewgdewar** [comment]
Below is a comparison between your current implementation and what is currently done on the "Customs repository" page:

Screenshot 2024-01-23 at 11 13 40 AM

Screenshot 2024-01-23 at 11 12 25 AM

Notice the padding/style differs quite a bit. 

Please adjust your implementation to match.

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx [hunk 9]
```diff
@@ -81,3 +81,25 @@ it('Render a loading state checking search disabled', () => {
...
+      <PopularRepositoriesTable />
+    </ReactQueryTestWrapper>,
+  );
+
+  const searchBox = getByPlaceholderText('Filter by name/url');
+  fireEvent.change(searchBox, { target: { value: 'yourSearchText' } });
+
+  waitFor(() => {
+    const occurrences = queryByText('yourSearchText', { exact: false });
+    expect(occurrences).toBeTruthy();
```
Comment (Andrewgdewar) [inline src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx:103]: This is starting to look good!  
But you need to find all "occurrences". In this case I believe you can use the "query**All**ByText". 

That method would return an array of items that are a match.

Currently if there is more than one item found you'd likely get an error that the test doesn't know which one you are wanting to target.

So once you have an array of items, you can check the length of that array (array.length) and have the expect look for a length of 2. 
There are a couple ways to do that with "expect" so I'll leave that for you to figure out.

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx [hunk 10]
```diff
@@ -81,3 +81,26 @@ it('Render a loading state checking search disabled', () => {
...
+    </ReactQueryTestWrapper>,
+  );
+
+  const searchBox = getByPlaceholderText('Filter by name/url');
+  fireEvent.change(searchBox, { target: { value: 'yourSearchText' } });
+
+  waitFor(() => {
+    const occurrences = queryAllByText('yourSearchText', { exact: false });
+    expect(occurrences).toBeTruthy();
+    expect(occurrences.length).toBe(2);
```
Comment (Andrewgdewar) [inline src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx:104]: Question, does this test still pass if you change this number to 1 or 4 or 88?

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx [hunk 11]
```diff
@@ -81,3 +81,26 @@ it('Render a loading state checking search disabled', () => {
...
+    <ReactQueryTestWrapper>
+      <PopularRepositoriesTable />
+    </ReactQueryTestWrapper>,
+  );
+
+  const searchBox = getByPlaceholderText('Filter by name/url');
+  fireEvent.change(searchBox, { target: { value: 'yourSearchText' } });
+
+  waitFor(() => {
+    const occurrences = queryAllByText('yourSearchText', { exact: false });
```
Comment (Andrewgdewar) [inline src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx:102]: if you change this "yourSearchText" to "somethingElse" will this test still pass?

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx [hunk 12]
```diff
@@ -81,3 +81,26 @@ it('Render a loading state checking search disabled', () => {
...
+    expect(occurrences).toBeTruthy();
```
Comment (Andrewgdewar) [inline src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx:103]: Read [this](https://developer.mozilla.org/en-US/docs/Glossary/Truthy) on what "truthy means" (it just means not negative value) 
Query all by text returns an empty array "[]". 
Screenshot 2024-01-24 at 11 50 04 AM

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx (hunk 11 above)
Comment (kwarnerredhat) [reply src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx:102]: Yeah it still passs

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx (hunk 10 above)
Comment (kwarnerredhat) [reply src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx:104]: Yeah it still pass

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx (hunk 12 above)
Comment (kwarnerredhat) [reply src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx:103]: It prints out an array in the boolean context and highlights if its true

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx [hunk 13]
```diff
@@ -81,3 +81,26 @@ it('Render a loading state checking search disabled', () => {
...
+  const { getByPlaceholderText, queryAllByText } = render(
+    <ReactQueryTestWrapper>
+      <PopularRepositoriesTable />
+    </ReactQueryTestWrapper>,
+  );
+
+  const searchBox = getByPlaceholderText('Filter by name/url');
+  fireEvent.change(searchBox, { target: { value: 'yourSearchText' } });
+
+  waitFor(() => {
```
Comment (Andrewgdewar) [inline src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx:101]: waitFor is only used with async methods, and in this case needs to "await" ed.

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx (hunk 12 above)
Comment (Andrewgdewar) [reply src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx]: Reread the above

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx [hunk 14]
```diff
@@ -81,3 +81,22 @@ it('Render a loading state checking search disabled', () => {
...
+  const { getByPlaceholderText, queryAllByText } = render(
+    <ReactQueryTestWrapper>
+      <PopularRepositoriesTable />
+    </ReactQueryTestWrapper>,
+  );
+
+  const searchBox = getByPlaceholderText('Filter by name/url');
+  fireEvent.change(searchBox, { target: { value: 'yourSearchText' } });
+
+  await waitFor(() => queryAllByText('yourSearchText', { exact: false }).length === 2);
```
Comment (Andrewgdewar) [inline src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx:101]: this doesn't test anything you need to use the "expect" methods in tests.

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx [hunk 15]
```diff
@@ -81,3 +81,28 @@ it('Render a loading state checking search disabled', () => {
 
   expect(queryByPlaceholderText('Filter by name/url')).toHaveAttribute('disabled');
 });
+
+it('finds search box, enters text, and checks text occurrence', async () => {
```
Comment (Andrewgdewar) [inline src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx:85]: this test does not need to be async

Comment (kwarnerredhat) [reply src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.test.tsx]: I removed it

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx [hunk 16]
```diff
@@ -125,6 +136,7 @@ const PopularRepositoriesTable = () => {
   const storedPerPage = Number(localStorage.getItem(perPageKey)) || 20;
   const [page, setPage] = useState(1);
   const [searchValue, setSearchValue] = useState('');
+  const [searchQuery, setSearchQuery] = useState('');
```
Comment (Andrewgdewar) [inline src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx:139]: We already have a value that stores the search text, it is called "searchValue" 
This extra state is not needed.

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx [hunk 17]
```diff
@@ -343,6 +355,11 @@ const PopularRepositoriesTable = () => {
 
   const countIsZero = !data?.data?.length;
 
+  const clearFilters = () => {
+    setSearchValue('');
+    setSearchQuery('');
```
Comment (Andrewgdewar) [inline src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx:360]: as mentioned above, this does the same thing.

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx [hunk 18]
```diff
@@ -343,6 +355,11 @@ const PopularRepositoriesTable = () => {
 
   const countIsZero = !data?.data?.length;
 
+  const clearFilters = () => {
```
Comment (Andrewgdewar) [inline src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx:358]: As this is just setting the searchValue to "". 
Do we need a function for this? or can we call it directly.

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx [hunk 19]
```diff
@@ -452,6 +468,25 @@ const PopularRepositoriesTable = () => {
               </FlexItem>
             </InputGroupItem>
           </InputGroup>
+          {(debouncedSearchValue !== '' || searchQuery !== '') && (
```
Comment (Andrewgdewar) [inline src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx:471]: Look into what debounced values are. 
Would it make sense to wait for the debouncedSearchValue to change this? 
Try typing something in the searchbox, and then hit clear filters. 
Does the chip immediately disappear?

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx [hunk 20]
```diff
@@ -355,12 +372,11 @@ const PopularRepositoriesTable = () => {
...
               <TextInput
                 isDisabled={isLoading}
-                type='text'
                 id='search'
-                ouiaId='popular_filter_search'
+                ouiaId='filter_search'
                 placeholder='Filter by name/url'
-                value={searchValue}
-                onChange={(_event, val) => setSearchValue(val)}
+                value={searchQuery}
```
Comment (Andrewgdewar) [inline src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx:378]: you've changed this value to a new one you've created, but the API is still using the old one. 

This makes it so that when you type into the search box, the api is not filtered. 

So at the moment, the searchbox does not function as intended.

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx (hunk 18 above)
Comment (kwarnerredhat) [reply src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx:358]: So the function might be needed

Comment (Andrewgdewar) [reply src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx:358]: It's not. if you remove searchQuery/setSearchQuery and use searchValue/setSearchValue. 
Then replace this function "clearFilters" with ()=> setSearchValue('') 
You will have the same functionality, and you will actually be able to "search"

File: src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx [hunk 21]
```diff
@@ -366,8 +366,9 @@ const PopularRepositoriesTable = () => {
             <InputGroupItem isFill>
               <TextInput
                 isDisabled={isLoading}
+                type='text'
                 id='search'
-                ouiaId='filter_search'
+                ouiaId='popular_filter_search'
```
Comment (Andrewgdewar) [inline src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx:371]: I add this back as QE uses this to target the input for tests. 
This is likely the reason that your CI-CD Build was failing.

Comment (kwarnerredhat) [reply src/Pages/PopularRepositoriesTable/PopularRepositoriesTable.tsx:371]: Okay

**Review by Andrewgdewar — APPROVED** [review]
Changes look. good.

**Comment by swadeley** [comment]
Hi

its looks good:

image


and it works.

Thank you.
